### DISCUSSION TITLE
Zustand with vitest spyOn

### DISCUSSION THREAD

**Discussion by shorthander** [pmndrs/zustand#2130] [Q&A] [answered]
Hi guys,
I am trying to test one of my hooks at the moment. The hook uses a Zustand store. I would like to check if one of the methods that is changing the store is getting called. I have no idea how to solve it... Really appreciate any suggestion how this could be done.

this is my  hook code: 

```
import { create } from 'zustand';

import { DataObject } from '@/src/features/data';

interface CreateDataModalStore {
  isOpen: boolean;
  onOpen: () => void;
  onClose: () => void;
  dataObject: DataObject;
  setDataObject: (data: DataObject) => void;
}

const deletedData: DataObject = {
data: ''
};

const useCreateDataModal = create<CreateDataModalStore>((set) => ({
  isOpen: false,
  dataObject: deletedData,
  onOpen: () => set({ isOpen: true }),
  onClose: () => {
    set({ isOpen: false });
    set({ dataObject: deletedData });
  },
  setDataObject: (data) => set({ dataObject: data }),
}));

export default useCreateDataModal;
```

This is the Component under test: 

```
const Header = () => {
  const t = useTranslations('Header');
  const { onOpen } = useCreateDataModal();
  const handleButton = () => {
    onOpen();
  };

  return (
    <div className={'flex flex-row justify-between'}>
      <h1 className="text-3xl font-bold">{t('data')}</h1>
      <Button onClick={handleButton}>{t('createNewData')}</Button>
    </div>
  );
};
```

And this is the test I am trying to execute:
```
describe('Data Page Header', () => {

  it('calls onOpen when the button is clicked', async () => {
    const spy = vi.spyOn(useCreateDataModal, 'onOpen');
    const screen = render(<Header />);
    const button = screen.getByText('createNewData');

    fireEvent.click(button);

    expect(spy).toHaveBeenCalledTimes(1);
  });

});
```

onOpen is not known and if I try to use getState vitest is complaining that it is getting called 0 times. I am a bit lost and have no idea how to test this thing. 
Thank you for your help :) :pray:

**Comment by dai-shi** [comment]
@dbritto-dev would probably be able to help it.

**Reply by dbritto-dev** [reply]
@dai-shi sure, @shorthander let me create an example for you.

**Reply by shorthander** [reply]
Thank you so much! really appreciate it :)

**Reply by dbritto-dev** [reply, marked as answer]
@shorthander here you go -> https://codesandbox.io/p/sandbox/zustand-vitest-demo-ph5gnj?file=%2Fsrc%2Fcomponents%2Fcounter%2Fcounter.test.tsx%3A31%2C3-41%2C6

In a nutshell you can use `vi.spyOn()` on `useAnyStore.getState()` function 

```tsx
test("should call inc func", async () => {
  const incSpy = vi.spyOn(useCounterStore.getState(), "inc");

  renderCounter();

  await act(async () => {
    await user.click(await screen.findByRole("button", { name: /one up/i }));
  });

  expect(incSpy).toHaveBeenCalled();
});
```

**Reply by shorthander** [reply]
Holy cow... Its so obvious and I wasted hours of my time. 
Thank you so much for your help!

**Reply by dbritto-dev** [reply]
@shorthander btw, would onOpen should display something? in that case I would test that out instead of the implementation itself

**Reply by shorthander** [reply]
Yeah you are right. Depends on the test level :) I was starting with unit testing so it was the right way to spy it. On integration test I agree with you. And to be honest I was setting up everything and picked the "easiest" component to test and ran in trouble with that. Now I know it! Thanks again

**Comment by taiweituan** [comment]
This has been a headache to figure out for using `spyOn` with Zustand + Context API due to needing SSR, but for the later visitor that stumbles upon this thread, here's my workaround (somewhat TypeScript friendly)  which combined `dbritto-dev`'s example -- 

```tsx
let contextSpy: ZStoreType

it('should assert called args', async () => {
  render(
    <ZStoreProvider
      initialState={{ // props from SSR
        userLocation: { locationId: '123', latLong: '45.6789, -122.3456' },
      }}
    >
      <ZStoreContext.Consumer>
        {(context) => {
          contextSpy = context as unknown as ZStoreType
          return (
            <OtherChildrenComponent>
              <MoarChildrenComponent>{children}</MoarChildrenComponent>
            </OtherChildrenComponent>
          )
        }}
      </ZStoreContext.Consumer>
    </ZStoreProvider>,
  )

  // Look here
  const useSetLocationIdTokenLocationActionSpy = jest.spyOn(contextSpy.getState(), 'setUserLocationAction')

  await userEvent.click(screen.getByTestId('set-location-button'))
  // ... more user events as needed

  expect(useSetLocationIdTokenLocationActionSpy).toHaveBeenCalledWith('some-location', 'chicago')
})

```

It's ugly, but it's the only way I could get it to work after a long time. Hopefully, there will be a tool or utility testing function in the future to make writing tests for this scenario easier.
